go/consensus/tendermint/apps/governance/transactions.go

```diff
@@ -0,0 +1,288 @@
+package governance
+
+import (
+	"fmt"
+
+	"github.com/oasisprotocol/oasis-core/go/common/cbor"
+	"github.com/oasisprotocol/oasis-core/go/common/node"
+	"github.com/oasisprotocol/oasis-core/go/consensus/tendermint/api"
+	governanceState "github.com/oasisprotocol/oasis-core/go/consensus/tendermint/apps/governance/state"
+	registryState "github.com/oasisprotocol/oasis-core/go/consensus/tendermint/apps/registry/state"
+	schedulerState "github.com/oasisprotocol/oasis-core/go/consensus/tendermint/apps/scheduler/state"
+	stakingState "github.com/oasisprotocol/oasis-core/go/consensus/tendermint/apps/staking/state"
+	governance "github.com/oasisprotocol/oasis-core/go/governance/api"
+	registryAPI "github.com/oasisprotocol/oasis-core/go/registry/api"
+	stakingAPI "github.com/oasisprotocol/oasis-core/go/staking/api"
+	upgradeAPI "github.com/oasisprotocol/oasis-core/go/upgrade/api"
+)
+
+func (app *governanceApplication) submitProposal(
+	ctx *api.Context,
+	state *governanceState.MutableState,
+	proposalContent *governance.ProposalContent,
+) error {
+	ctx.Logger().Debug("governance: submit proposal tx",
+		"proposal_content", proposalContent,
+	)
+
+	params, err := state.ConsensusParameters(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to fetch consensus parameters: %w", err)
+	}
+
+	// Validate proposal content basics.
+	if err = proposalContent.ValidateBasic(); err != nil {
+		ctx.Logger().Error("governance: malformed proposal content",
+			"content", proposalContent,
+			"err", err,
+		)
+		return governance.ErrInvalidArgument
+	}
+
+	if ctx.IsCheckOnly() {
```

Comment:
I moved the `IsCheckOnly()` check for both `submitProposal` and `castVote` methods.  Please take another quick look over this part @kostko 

Actually could probably move the check in `submitProposal` even further bellow probably, not sure how much does it make sense though.

Reply:
I would move fetching consensus parameters below (as they are not used in CheckTx and avoidng stateful operations is a good thing), otherwise this looks good.

Reply:
done